lib/api/src/lib.rs

```diff
@@ -118,3 +120,118 @@ pub use wasmer_engine_native::{Native, NativeArtifact, NativeEngine};
 
 /// Version number of this crate.
 pub const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+// --------------------------------------------------------------
+// TODO: put this in a proper location, just prototyping for now:
+// TODO: rename everything, all names are throw-away names
+
+/// Prototype trait for finishing envs.
+pub trait WasmerEnv {
+    /// The function that Wasmer will call on your type to let it finish
+    /// instantiating.
+    fn finish(&mut self, instance: &Instance);
+
+    /// Frees memory written to `self` so it can be dropped without any memory leaks.
+    fn free(&mut self);
+}
+
+impl<T: WasmerEnv> WasmerEnv for &'static mut T {
+    fn finish(&mut self, instance: &Instance) {
+        (*self).finish(instance)
+    }
+    fn free(&mut self) {
+        (*self).free()
+    }
+}
+
+// TODO: iterate on names
+// TODO: do we want to use mutex/atomics here? like old WASI solution
+/// Lazily init an item
+pub struct InitAfterInstance<T: Sized> {
+    /// The data to be initialized
+    data: std::mem::MaybeUninit<T>,
+    /// Whether or not the data has been initialized
+    initialized: bool,
+}
+
+impl<T> InitAfterInstance<T> {
+    /// Creates an unitialized value.
+    pub fn new() -> Self {
+        Self {
+            data: std::mem::MaybeUninit::uninit(),
+            initialized: false,
+        }
+    }
+
+    /// # Safety
+    /// - The data must be initialized first
+    pub unsafe fn get_unchecked(&self) -> &T {
+        &*self.data.as_ptr()
+    }
+
+    /// Get the inner data.
+    pub fn get_ref(&self) -> Option<&T> {
+        if !self.initialized {
+            None
+        } else {
+            Some(unsafe { self.get_unchecked() })
+        }
+    }
+
+    /// TOOD: review
+    pub fn initialize(&mut self, value: T) -> bool {
+        if self.initialized {
+            return false;
+        }
+        unsafe {
+            self.data.as_mut_ptr().write(value);
+        }
+        self.initialized = true;
+        true
+    }
+}
+
+impl<T: std::fmt::Debug> std::fmt::Debug for InitAfterInstance<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.debug_struct("InitAfterInstance")
+            .field("data", &self.get_ref())
+            .finish()
+    }
+}
+
+impl<T: Clone> Clone for InitAfterInstance<T> {
+    fn clone(&self) -> Self {
+        if let Some(inner) = self.get_ref() {
+            Self {
+                data: std::mem::MaybeUninit::new(inner.clone()),
+                initialized: true,
+            }
+        } else {
+            Self {
+                data: std::mem::MaybeUninit::uninit(),
+                initialized: false,
+            }
+        }
+    }
+}
```

Comment:
Why do we want `InitAfterInstance` to be clonable? Is it to preserve the behavior of `T` in this case (because `T` is clonable, then `InitAfterInstance<T>` is also clonable)?

Reply:
Hmm, on second thought I'm not exactly sure why I felt the need to add this, I think it's just because `WasiEnv` was clone and it let me keep `derive(Clone)` on it

Reply:
Probably we don't need WasiEnv to be clonable after this PR?